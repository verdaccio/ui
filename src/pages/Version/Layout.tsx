import React, { FC, ReactElement } from 'react';
import Grid from '@material-ui/core/Grid';
import DetailContainer from '../../components/DetailContainer';
import DetailSidebar from '../../components/DetailSidebar';

function renderDetail(): ReactElement<HTMLElement> {
  return <DetailContainer />;
}

function renderSidebar(): ReactElement<HTMLElement> {
  return <DetailSidebar />;
}

const Layout: FC<{}> = () => {
  return (
    <Grid className={'container content'} container={true} data-testid="version-layout" spacing={0}>
      <Grid item={true} md={8} xs={12}>
        {renderDetail()}
      </Grid>
      <Grid item={true} md={4} xs={12}>
        {renderSidebar()}
      </Grid>
    </Grid>
  );
};

export { Layout };
