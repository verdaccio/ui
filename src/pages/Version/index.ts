export { DetailContext, DetailContextConsumer, DetailContextProvider } from './context';
export { VersionPageConsumerProps, DetailContextProps } from './version-config';

export { default } from './Version';
