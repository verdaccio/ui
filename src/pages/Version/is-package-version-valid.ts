import { PackageMetaInterface } from '../../../types/packageMeta';

function isPackageVersionValid(packageMeta: Partial<PackageMetaInterface>, packageVersion?: string): boolean {
  if (!packageVersion || typeof packageVersion === 'undefined') {
    // if is undefined, that means versions does not exist, we continue
    return true;
  }

  if (packageMeta.versions) {
    return Object.keys(packageMeta.versions).includes(packageVersion);
  }

  return false;
}

export default isPackageVersionValid;
