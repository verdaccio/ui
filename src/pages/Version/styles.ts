import styled from 'react-emotion';
import DialogTitle from '@material-ui/core/DialogTitle';

import colors from '../../utils/styles/colors';
import { fontSize } from '../../utils/styles/sizes';

export const Title = styled(DialogTitle)({
  '&&': {
    backgroundColor: colors.primary,
    color: colors.white,
    fontSize: fontSize.lg,
  },
});
