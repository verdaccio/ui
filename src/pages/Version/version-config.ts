import { PackageMetaInterface } from '../../../types/packageMeta';

export interface DetailContextProps {
  enableLoading: () => void;
  hasNotBeenFound: boolean;
  isLoading: boolean;
  packageMeta: PackageMetaInterface;
  packageName: string;
  packageVersion?: string;
  readMe: string;
}

export interface VersionPageConsumerProps {
  enableLoading: () => void;
  packageMeta: PackageMetaInterface;
  packageName: string;
  packageVersion?: string;
  readMe: string;
}
