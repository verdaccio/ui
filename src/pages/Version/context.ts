import { createContext, Consumer, Provider } from 'react';

import { DetailContextProps, VersionPageConsumerProps } from './version-config';

export const DetailContext = createContext<Partial<DetailContextProps>>({});

export const DetailContextProvider: Provider<Partial<VersionPageConsumerProps>> = DetailContext.Provider;
export const DetailContextConsumer: Consumer<Partial<VersionPageConsumerProps>> = DetailContext.Consumer;
