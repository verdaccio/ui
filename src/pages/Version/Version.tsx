import React, { useContext } from 'react';

import Loading from 'verdaccio-ui/components/Loading';
import NotFound from 'verdaccio-ui/components/NotFound';

import { DetailContext } from './context';
import VersionLayout from './VersionLayout';

const Version: React.FC = () => {
  const detailContext = useContext(DetailContext);
  const { isLoading, hasNotBeenFound } = detailContext;

  if (isLoading) {
    return <Loading />;
  }

  if (hasNotBeenFound) {
    return <NotFound />;
  }

  return <VersionLayout />;
};

export default Version;
