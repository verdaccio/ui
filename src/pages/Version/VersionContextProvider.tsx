import React, { useEffect, useState, useCallback } from 'react';
import { useParams } from 'react-router-dom';
import { PackageMetaInterface } from 'types/packageMeta';

import { useAPI } from 'verdaccio-ui/providers/API/APIProvider';

import { DetailContext } from './context';
import getRouterPackageName from './get-route-package-name';
import isPackageVersionValid from './is-package-version-valid';

interface Params {
  scope?: string;
  package: string;
  version?: string;
}

const VersionContextProvider: React.FC = ({ children }) => {
  const { version, package: pkgName, scope } = useParams<Params>();
  const [packageName, setPackageName] = useState(getRouterPackageName(pkgName, scope));
  const [packageVersion, setPackageVersion] = useState(version);
  const [packageMeta, setPackageMeta] = useState<PackageMetaInterface>();
  const [readMe, setReadme] = useState<string>();
  const [isLoading, setIsLoading] = useState<boolean>(true);
  const [hasNotBeenFound, setHasNotBeenFound] = useState<boolean>();
  const { callDetailPage, callReadme } = useAPI();

  useEffect(() => {
    const updatedPackageName = getRouterPackageName(pkgName, scope);
    setPackageName(updatedPackageName);
  }, [pkgName, scope]);

  useEffect(() => {
    setPackageVersion(version);
  }, [version]);

  const doCalls = useCallback(async () => {
    try {
      const packageMeta = await callDetailPage(packageName, packageVersion);
      const readMe = await callReadme(packageName, packageVersion);
      if (isPackageVersionValid(packageMeta, packageVersion)) {
        setReadme(readMe);
        setPackageMeta(packageMeta);
        setIsLoading(false);
      } else {
        setIsLoading(false);
        setHasNotBeenFound(true);
      }
    } catch (error) {
      setHasNotBeenFound(true);
      setIsLoading(false);
    }
  }, [packageName, packageVersion, callDetailPage, callReadme]);

  useEffect(() => {
    doCalls();
  }, [doCalls]);

  return (
    <DetailContext.Provider
      value={{
        packageMeta,
        packageVersion,
        readMe,
        packageName,
        isLoading,
        hasNotBeenFound,
      }}>
      {children}
    </DetailContext.Provider>
  );
};

export default VersionContextProvider;
