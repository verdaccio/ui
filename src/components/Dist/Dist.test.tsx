import React from 'react';
import { mount } from 'enzyme';

import { DetailContext } from '../../pages/Version';
import Dist from './Dist';

const withDistComponent = (packageMeta: React.ContextType<typeof DetailContext>['packageMeta']): JSX.Element => (
  <DetailContext.Provider value={{ packageMeta }}>
    <Dist />
  </DetailContext.Provider>
);

describe('<Dist /> component', () => {
  test('should render the component in default state', () => {
    const packageMeta = {
      latest: {
        name: 'verdaccio',
        version: '4.0.0',
        dist: {
          fileCount: 7,
          unpackedSize: 10,
        },
        license: '',
      },
      _uplinks: {},
    };

    const wrapper = mount(withDistComponent(packageMeta));
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render the component with license as string', () => {
    const packageMeta = {
      latest: {
        name: 'verdaccio',
        version: '4.0.0',
        dist: {
          fileCount: 7,
          unpackedSize: 10,
        },
        license: 'MIT',
      },
      _uplinks: {},
    };

    const wrapper = mount(withDistComponent(packageMeta));
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('should render the component with license as object', () => {
    const packageMeta = {
      latest: {
        name: 'verdaccio',
        version: '4.0.0',
        dist: {
          fileCount: 7,
          unpackedSize: 10,
        },
        license: {
          type: 'MIT',
          url: 'https://www.opensource.org/licenses/mit-license.php',
        },
      },
      _uplinks: {},
    };

    const wrapper = mount(withDistComponent(packageMeta));
    expect(wrapper.html()).toMatchSnapshot();
  });
});
