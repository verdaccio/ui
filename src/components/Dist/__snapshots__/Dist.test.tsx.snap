// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dist /> component should render the component in default state 1`] = `"<ul class=\\"MuiList-root MuiList-padding MuiList-subheader\\"><h6 class=\\"MuiTypography-root css-1na337r-StyledText estxrtg0 MuiTypography-subtitle1\\">Latest Distribution</h6><div class=\\"MuiButtonBase-root MuiListItem-root css-1mms18p-DistListItem estxrtg1 MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\"><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>file count</b>: 7</span></div><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>size</b>: 10.00 Bytes</span></div><span class=\\"MuiTouchRipple-root\\"></span></div></ul>"`;

exports[`<Dist /> component should render the component with license as object 1`] = `"<ul class=\\"MuiList-root MuiList-padding MuiList-subheader\\"><h6 class=\\"MuiTypography-root css-1na337r-StyledText estxrtg0 MuiTypography-subtitle1\\">Latest Distribution</h6><div class=\\"MuiButtonBase-root MuiListItem-root css-1mms18p-DistListItem estxrtg1 MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\"><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>file count</b>: 7</span></div><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>size</b>: 10.00 Bytes</span></div><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>license</b>: MIT</span></div><span class=\\"MuiTouchRipple-root\\"></span></div></ul>"`;

exports[`<Dist /> component should render the component with license as string 1`] = `"<ul class=\\"MuiList-root MuiList-padding MuiList-subheader\\"><h6 class=\\"MuiTypography-root css-1na337r-StyledText estxrtg0 MuiTypography-subtitle1\\">Latest Distribution</h6><div class=\\"MuiButtonBase-root MuiListItem-root css-1mms18p-DistListItem estxrtg1 MuiListItem-gutters MuiListItem-button\\" tabindex=\\"0\\" role=\\"button\\" aria-disabled=\\"false\\"><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>file count</b>: 7</span></div><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>size</b>: 10.00 Bytes</span></div><div class=\\"MuiChip-root css-e2le7v-DistChips estxrtg2\\"><span class=\\"MuiChip-label\\"><b>license</b>: MIT</span></div><span class=\\"MuiTouchRipple-root\\"></span></div></ul>"`;
